src/Everscale/CellBuilder.cpp

```diff
@@ -0,0 +1,268 @@
+    appendRaw(appendedData, 8);
+}
+
+void CellBuilder::appendBits(uint64_t value, uint8_t bits) {
+    Data appendedData;
+
+    if (bits == 0) {
+        // Do nothing
+    } else if (bits >= 1 && bits <= 7) {
+        auto val = static_cast<uint8_t>(value << (8 - bits));
+        appendedData.push_back(val);
+    } else if (bits >= 8 && bits <= 15) {
+        auto val = static_cast<uint16_t>(value << (16 - bits));
+        encode16BE(val, appendedData);
+    } else if (bits >= 16 && bits <= 31) {
+        auto val = static_cast<uint32_t>(value << (32 - bits));
+        encode32BE(val, appendedData);
+    } else if (bits >= 32 && bits <= 63) {
+        auto val = static_cast<uint64_t>(value << (64 - bits));
+        encode64BE(val, appendedData);
+    }
+
+    appendRaw(appendedData, bits);
+}
+
+void CellBuilder::appendRaw(const Data& appendedData, uint16_t bits) {
+    if (appendedData.size() * 8 < bits) {
+        throw std::invalid_argument("invalid builder data");
+    } else if (bitLen + bits > Cell::MAX_BITS) {
+        throw std::runtime_error("cell data overflow");
+    } else if (bits != 0) {
+        if ((bitLen % 8) == 0) {
+            if ((bits % 8) == 0) {
+                appendWithoutShifting(appendedData, bits);
+            } else {
+                appendWithSliceShifting(appendedData, bits);
+            }
+        } else {
+            appendWithDoubleShifting(appendedData, bits);
+        }
+    }
+    assert(bitLen <= Cell::MAX_BITS);
+    assert(data.size() * 8 <= Cell::MAX_BITS + 1);
+}
+
+void CellBuilder::prependRaw(Data& appendedData, uint16_t bits) {
+    if (bits != 0) {
+        auto buffer = CellBuilder(appendedData, bits);
+        buffer.appendRaw(data, bitLen);
+
+        data = buffer.data;
+        bitLen = buffer.bitLen;
+    }
+}
+
+void CellBuilder::appendReferenceCell(std::shared_ptr<Cell> child) {
+    if (references.size() + 1 > Cell::MAX_REFS) {
+        throw std::runtime_error("cell refs overflow");
+    }
+    references.push_back(std::move(child));
+}
+
+void CellBuilder::appendBuilder(const CellBuilder& builder) {
+    appendRaw(builder.data, builder.bitLen);
+    for (const auto & reference : builder.references) {
+        appendReferenceCell(reference);
+    }
+}
+
+void CellBuilder::appendCellSlice(const CellSlice &other) {
+    Data appendedData(other.cell->data);
+    appendRaw(appendedData, other.cell->bitLen);
+
+    for (auto i = 0; i < other.cell->refCount; i++) {
+        appendReferenceCell(other.cell->references[i]);
+    }
+}
+
+Cell::Ref CellBuilder::intoCell() {
+    // Append tag
+    if (bitLen & 7) {
+        const auto mask = static_cast<uint8_t>(0x80 >> (bitLen & 7));
+        const auto l = bitLen / 8;
+        data[l] = static_cast<uint8_t>((data[l] & ~mask) | mask);
+    }
+
+    auto refCount = references.size();
+
+    Cell::Refs refs;
+    for (size_t r = 0; r < refCount; ++r) {
+        refs[r] = std::move(references[r]);
+    }
```

Comment:
You probably want to use `https://en.cppreference.com/w/cpp/algorithm/move` instead

Reply:
Done